### ISSUE TITLE
seg fault if NIST EC public key not supplied as param

### ISSUE THREAD

**Issue by sftcd** [openssl/openssl#18797]
Please remember to tell us in what OpenSSL version you found the issue.

- bug is in master  branch

Context is creating an EC key pair starting from the secret key. That works fine for x25519/x448 but can cause a segfault for NIST  curves If one omits a call to add the public key from outside via e.g.  ``OSSL_PARAM_BLD_push_octet_string(param_bld, "pub", pubuf, pubuf_len)``. End result is that one has to re-calculate the public key outside the library (which isn't ideal but is ok), but the crash seems like a bad thing. I don't currently see that as a vulnerability though, just a bug.

The problem seems to be a NULL pointer de-ref within the EC handling code. My workaround is below but likely isn't the correct fix.

```diff --git a/providers/implementations/keymgmt/ec_kmgmt.c b/providers/implementations/keymgmt/ec_kmgmt.c
index 7aed057cac..da274c57d9 100644
--- a/providers/implementations/keymgmt/ec_kmgmt.c
+++ b/providers/implementations/keymgmt/ec_kmgmt.c
@@ -726,7 +726,8 @@ int common_get_params(void *key, OSSL_PARAM params[], int sm2)
         }
     }
     if ((p = OSSL_PARAM_locate(params,
-                               OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY)) != NULL) {
+                               OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY)) != NULL &&
+        EC_KEY_get0_public_key(key) != NULL) {
         p->return_size = EC_POINT_point2oct(EC_KEY_get0_group(key),
                                             EC_KEY_get0_public_key(key),
                                             POINT_CONVERSION_UNCOMPRESSED,
```

POC code causing a crash is [here](https://github.com/sftcd/happykey/blob/master/test2evp.c). Stack trace from the crash is below:

```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7b9faa0 in ec_point_is_compat (point=0x0, group=0x555555583db0) at crypto/ec/ec_local.h:328
328	    return group->meth == point->meth
(gdb) bt
#0  0x00007ffff7b9faa0 in ec_point_is_compat (point=0x0, group=0x555555583db0) at crypto/ec/ec_local.h:328
#1  0x00007ffff7b9fd6d in EC_POINT_point2oct (group=0x555555583db0, point=0x0, form=POINT_CONVERSION_UNCOMPRESSED, buf=0x0, len=18446744073709551615, ctx=0x5555555848a0)
    at crypto/ec/ec_oct.c:82
#2  0x00007ffff7e4bac4 in common_get_params (key=0x555555583c10, params=0x7fffffffdbc0, sm2=0) at providers/implementations/keymgmt/ec_kmgmt.c:725
#3  0x00007ffff7e4bc3a in ec_get_params (key=0x555555583c10, params=0x7fffffffdbc0) at providers/implementations/keymgmt/ec_kmgmt.c:749
#4  0x00007ffff7c2a9bc in evp_keymgmt_get_params (keymgmt=0x555555580230, keydata=0x555555583c10, params=0x7fffffffdbc0) at crypto/evp/keymgmt_meth.c:392
#5  0x00007ffff7c36b54 in EVP_PKEY_get_params (pkey=0x555555583b30, params=0x7fffffffdbc0) at crypto/evp/p_lib.c:2346
#6  0x00007ffff7c35fc0 in EVP_PKEY_get_octet_string_param (pkey=0x555555583b30, key_name=0x7ffff7ea75a5 "encoded-pub-key", buf=0x0, max_buf_sz=0, out_len=0x7fffffffdc50)
    at crypto/evp/p_lib.c:2166
#7  0x00007ffff7c343c9 in EVP_PKEY_get1_encoded_public_key (pkey=0x555555583b30, ppub=0x7fffffffdcb8) at crypto/evp/p_lib.c:1394
#8  0x000055555555730b in main (argc=1, argv=0x7fffffffde28) at test2evp.c:197
```

It'd also seem like a reasonable plan to add a new API allowing easier construction of a key pair from just the secret. I have code like that [here](https://github.com/sftcd/happykey/blob/master/hpke.c#L1680) but it probably needs a bit more work before one would want to add that as a new API. There's an API as part of my [HPKE PR](https://github.com/openssl/openssl/pull/17172) that will contain such code. (I've yet to push that to the HPKE-PR branch, will do tomorrow or so.)

**Comment by t8m** [comment]
EC keys without a public key are supposed to work, so yeah this is clearly a bug. The question is whether we should compute the public key in such case. Or whether we should have a function to compute it.

**Comment by t8m** [comment]
At minimum the call should fail and not segfault.

**Comment by sftcd** [comment]
Hiya,

On 14/07/2022 07:59, Tomáš Mráz wrote:
> EC keys without a public key are supposed to work, so yeah this is
> clearly a bug. The question is whether we should compute the public
> key in such case. Or whether we should have a function to compute
> it.

I'd argue the former is a bit better, but if the latter were
done, then it should be trivial to also do the former inside
the library too, so that'd be my suggested approach. OTOH, I
guess as this hasn't been seen 'till now, not so many people
need the functionality:-)

Cheers,
S.

**Comment by t8m** [comment]
The problem is that computing the public key can in theory make a timing or other side channel for an attacker. So if the public key is computed when the private key is being loaded by the application, it is better than computing it only at the time when you need it i.e. when the peer asked for it.

**Comment by sftcd** [comment]
On 14/07/2022 12:55, Tomáš Mráz wrote:
> The problem is that computing the public key can in theory make a
> timing or other side channel for an attacker. So if the public key is
> computed when the private key is being loaded by the application, it
> is better than computing it only at the time when you need it i.e.
> when the peer asked for it.

Fair point. FWIW, I think I'd also prefer doing the
public key calculation automatically when only the
private is provided as input.

S

**Comment by skyboi1233** [comment]
> 
> 
> Please remember to tell us in what OpenSSL version you found the issue.
> 
> 
> 
> - bug is in master  branch
> 
> 
> 
> Context is creating an EC key pair starting from the secret key. That works fine for x25519/x448 but can cause a segfault for NIST  curves If one omits a call to add the public key from outside via e.g.  ``OSSL_PARAM_BLD_push_octet_string(param_bld, "pub", pubuf, pubuf_len)``. End result is that one has to re-calculate the public key outside the library (which isn't ideal but is ok), but the crash seems like a bad thing. I don't currently see that as a vulnerability though, just a bug.
> 
> 
> 
> The problem seems to be a NULL pointer de-ref within the EC handling code. My workaround is below but likely isn't the correct fix.
> 
> 
> 
> ```diff --git a/providers/implementations/keymgmt/ec_kmgmt.c b/providers/implementations/keymgmt/ec_kmgmt.c
> 
> index 7aed057cac..da274c57d9 100644
> 
> --- a/providers/implementations/keymgmt/ec_kmgmt.c
> 
> +++ b/providers/implementations/keymgmt/ec_kmgmt.c
> 
> @@ -726,7 +726,8 @@ int common_get_params(void *key, OSSL_PARAM params[], int sm2)
> 
>          }
> 
>      }
> 
>      if ((p = OSSL_PARAM_locate(params,
> 
> -                               OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY)) != NULL) {
> 
> +                               OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY)) != NULL &&
> 
> +        EC_KEY_get0_public_key(key) != NULL) {
> 
>          p->return_size = EC_POINT_point2oct(EC_KEY_get0_group(key),
> 
>                                              EC_KEY_get0_public_key(key),
> 
>                                              POINT_CONVERSION_UNCOMPRESSED,
> 
> ```
> 
> 
> 
> POC code causing a crash is [here](https://github.com/sftcd/happykey/blob/master/test2evp.c). Stack trace from the crash is below:
> 
> 
> 
> ```
> 
> Program received signal SIGSEGV, Segmentation fault.
> 
> 0x00007ffff7b9faa0 in ec_point_is_compat (point=0x0, group=0x555555583db0) at crypto/ec/ec_local.h:328
> 
> 328	    return group->meth == point->meth
> 
> (gdb) bt
> 
> #0  0x00007ffff7b9faa0 in ec_point_is_compat (point=0x0, group=0x555555583db0) at crypto/ec/ec_local.h:328
> 
> #1  0x00007ffff7b9fd6d in EC_POINT_point2oct (group=0x555555583db0, point=0x0, form=POINT_CONVERSION_UNCOMPRESSED, buf=0x0, len=18446744073709551615, ctx=0x5555555848a0)
> 
>     at crypto/ec/ec_oct.c:82
> 
> #2  0x00007ffff7e4bac4 in common_get_params (key=0x555555583c10, params=0x7fffffffdbc0, sm2=0) at providers/implementations/keymgmt/ec_kmgmt.c:725
> 
> #3  0x00007ffff7e4bc3a in ec_get_params (key=0x555555583c10, params=0x7fffffffdbc0) at providers/implementations/keymgmt/ec_kmgmt.c:749
> 
> #4  0x00007ffff7c2a9bc in evp_keymgmt_get_params (keymgmt=0x555555580230, keydata=0x555555583c10, params=0x7fffffffdbc0) at crypto/evp/keymgmt_meth.c:392
> 
> #5  0x00007ffff7c36b54 in EVP_PKEY_get_params (pkey=0x555555583b30, params=0x7fffffffdbc0) at crypto/evp/p_lib.c:2346
> 
> #6  0x00007ffff7c35fc0 in EVP_PKEY_get_octet_string_param (pkey=0x555555583b30, key_name=0x7ffff7ea75a5 "encoded-pub-key", buf=0x0, max_buf_sz=0, out_len=0x7fffffffdc50)
> 
>     at crypto/evp/p_lib.c:2166
> 
> #7  0x00007ffff7c343c9 in EVP_PKEY_get1_encoded_public_key (pkey=0x555555583b30, ppub=0x7fffffffdcb8) at crypto/evp/p_lib.c:1394
> 
> #8  0x000055555555730b in main (argc=1, argv=0x7fffffffde28) at test2evp.c:197
> 
> ```
> 
> 
> 
> It'd also seem like a reasonable plan to add a new API allowing easier construction of a key pair from just the secret. I have code like that [here](https://github.com/sftcd/happykey/blob/master/hpke.c#L1680) but it probably needs a bit more work before one would want to add that as a new API. There's an API as part of my [HPKE PR](https://github.com/openssl/openssl/pull/17172) that will contain such code. (I've yet to push that to the HPKE-PR branch, will do tomorrow or so.)
> 
>

**Comment by t8m** [comment]
This is basically a duplicate of #18495 which is fixed now by #18902. Closing this one as well.
